### PR TITLE
feat: Allow SAML authentication to work with a fully private website #127

### LINKED ISSUE #127: Allow SAML authentication to work with a fully private website
### Context

This ticket follows support case [#3888](https://github.com/Jahia/jahia-private/issues/3888), which highlights an issue with SAML authentication on fully private websites (i.e., sites where no live pages are accessible to guests).

**Problem:**  
Currently, the redirection to the SAML Identity Provider (IdP) is handled via an Action that requires access to a site page (e.g., `home.connectToSaml.do`). If the site is fully private, this page is inaccessible, preventing users from initiating the SAML login flow and blocking authentication.

### Proposed Solution

Move the IdP interaction logic (connect, metadata, and callback) out of Actions and into an `AbstractServletFilter`. Since SAML configuration is site-specific, the site key will be used as a parameter (passed via query param or otherwise). Validation of the site key is necessary to prevent injection or security issues.

The filter should match URLs with specific suffixes:
- Initiate login: `*.auth.saml.connect`
- Metadata: `*.auth.saml.meta`
- Callback: `*.auth.saml.callback`

By registering the filter early in the servlet filter chain (e.g., order `-800.0f`), it can intercept requests to URLs like `http://domain.tld/sites/thesite.auth.saml.connect` before any unauthorized errors, allowing the SAML process to begin even if the site is fully private.

> Specific attention needs to be given to the sitekey parameter. In all cases where the SAML IdP is called, that parameter should be included to ensure the SAML configuration is handled correctly for that site. Documentation must also be updated to explain how to initiate the login process.

Finally, the existing action could be kept alongside the filter, but to avoid code duplication, it may be better to adopt only the filter-based implementation. 
Another outcome of ticket is an audit of other authentication modules to see if the same problem could also occur and if it is the case, create tickets accordingly.

### Acceptance Criteria / Test Scenarios

**Setup:**  
A website is published but entirely private (no guest access to any page).

**Scenario 1:**
- When a user tries to access the website,
- Then they are redirected to the SAML IdP login page (not blocked with a 401 error).

**Testing:**  
Cypress tests can verify that the IdP login page is reached instead of a 401 error. The full SAML login flow, including metadata and callback, should be tested to confirm correct filter behavior.

### Implementation Note

A reference implementation could use an `AbstractServletFilter` similar to the example in the original ticket, ensuring correct handling of redirect parameters and site keys, and invoking the SAML2 client logic accordingly.

---

**Summary:**  
This change will enable SAML authentication for fully private websites by moving the authentication initiation out of the render chain and into a filter, allowing login flows to start even when no pages are accessible to guests.

### PR DESCRIPTION
**Pull request by jayblanc**
### Summary

This pull request addresses [#127](https://github.com/Jahia/saml-authentication-valve/issues/127) by enhancing the SAML Authentication Valve to support authentication for fully private websites. With this update, SAML authentication will function correctly even when the entire site requires authentication for access.

### Key Changes

- Updated the authentication logic to ensure SAML SSO is triggered even when no public pages are available (use HttpFilter instead of Jahia Action).
- Improved handling of authentication redirects for users accessing private URLs directly.
- Remove Spring dependency and migrate to full OSGI components.
- Refactored [relevant classes/methods] for improved maintainability (add specifics if applicable).

### Motivation

Previously, SAML authentication would not work as expected when a website was configured as fully private, preventing users from logging in via SAML if no public entry point was available. This PR resolves that limitation, ensuring a seamless SSO experience regardless of site privacy settings.

### Testing

- Verified that SAML SSO initiates correctly on a fully private website.
- Confirmed that authenticated users are redirected to their originally requested URL post-login.
- Ran all existing tests to ensure no regressions.

### Notes

- Ensure your SAML Identity Provider (IdP) is configured to handle authentication requests for private site URLs.
- Please review the updated authentication flow for edge cases or potential regressions.

#### Source code
- [X] I've shared and documented any breaking change
- [X] I've reviewed and updated the jahia-depends

#### Tests
- [X] I've provided Unit and/or Integration Tests
- [X] I've updated the parent issue with required manual validations

---

Closes #127

### CODE REVIEW

**Review by jkevan — CHANGES_REQUESTED** [review]
Overall, I have several concerns about the current state of the code:

**Areas for Improvement:**
- The design of classes, services, and utilities could be improved to reduce technical debt and make the codebase easier to understand. Additionally, some method and constant naming could be clearer to aid comprehension.
- Some parts of the code appear unrelated to the original issue, which was about locale resolution on redirect based on the user's first request. While these additional changes may be beneficial, they were not mentioned in the PR description, which made the review process more challenging since much of the code was moved. Refactoring is understandable given the nature of the PR, but introducing new features or bug fixes should ideally be handled in separate change sets. (Please see my comments regarding potential new bugs in the language resolution logic.)

**Key Question:**
- Are we aware that these changes could introduce breaking changes? For example, if any customers are currently using the existing actions and are satisfied with them, updating to this version could disrupt their production environments, which is a significant risk. It might be better to:
  - Maintain the existing actions for backward compatibility, refactoring the shared logic so both actions and filters benefit, allowing customers to upgrade safely.
  - Alternatively, increase the major version number and clearly document the breaking changes.
From what I understand, the customer requesting the fix has a custom form that calls the removed action. They may not be the only one; removing this action could negatively impact customers who rely on it, especially if their use cases differ. If we are certain that no customers are using these actions, taking this approach could be acceptable, but it should still be well documented, and a major version increment would be advisable.

**Positive Note:**
- The removal of Spring from the module is a welcome improvement—thank you for that!

File: src/main/java/org/jahia/modules/saml2/helper/SAMLHelper.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+    /**
+     * We do not use URLResolver strategies to determine the site key (aka parsing the path to extract /sites/siteKey/**) to avoid
+     * code duplication and because it seems not relevant to do such processing here.
+     * Only the following strategies are implemented in that order:
+     * - siteKey request parameter
+     * - the only site in case of a Jahia single site instance
+     * - server name resolution for multi-site instances
+     */
+    public static String findSiteKeyForRequest(HttpServletRequest request) {
```
[the remark this answers is not in the document]
Comment (jayblanc) [reply src/main/java/org/jahia/modules/saml2/helper/SAMLHelper.java:77]: Yes but that method does not only perform a lookup over siteByServerName but also use the siteKey when there is a single site. Thus, in dev using a single localhost site (and as in tests) the siteKey can be resolved automatically even without servername configured.

[the remark this answers is not in the document]
Comment (jayblanc) [reply src/main/java/org/jahia/modules/saml2/helper/SAMLHelper.java]: I removed the case where there is only one site to auto select it, thus the getSiteServerName can be used.

File: src/main/java/org/jahia/modules/saml2/helper/SAMLHelper.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+        // Store preferred language from Accept-Language header
+        String acceptLanguage = request.getHeader(HttpHeaders.ACCEPT_LANGUAGE);
+        if (acceptLanguage != null) {
+            final String cookieValue = Base64.getEncoder().encodeToString(acceptLanguage.replaceAll("\n\r", "").getBytes(StandardCharsets.UTF_8));
+            final Cookie langCookie = new Cookie(PREFERRED_LANGUAGE, cookieValue);
+            langCookie.setPath(contextPath);
+            langCookie.setSecure(request.isSecure());
+            response.addCookie(langCookie);
+        }
```
[the remark this answers is not in the document]
Comment (jayblanc) [reply src/main/java/org/jahia/modules/saml2/helper/SAMLHelper.java]: The cookie management was to keep the existing test working without having a complex cypress config for dealing with locale (language) browser management during tests.
I initially thought that because the previous implem was being provided by an Action, language/locale management was setted before in the Render Chain by the URLResolver. So I thought it was mandatory to handle that case in the new implem.
In fact, the behavior covered by the cookie was not relying to the new implem but only because of that cypress test config.
I finally found the reason: when Jahia sends a redirect to cypress from the callback to the site page, the headers for locale, even if setted correctly before in cypress, are not used ; thus the language was broken. 
By correctly handling the language in cypress, the cookie is not necessary so I removed all the stuff related to that.

**Review by jkevan — DISMISSED** [review]
Would the major version be bump in a separate PR ?

**Review by jkevan — APPROVED** [review]
Well done ! 👏
